### PR TITLE
Clean EmitDynamicCall code 

### PR DESCRIPTION
**Pull request by shargon**
- Set `CallFlags` based on snapshot

### CODE REVIEW

File: src/neo/VM/Helper.cs [hunk 1]
```diff
@@ -46,36 +48,41 @@ public static ScriptBuilder Emit(this ScriptBuilder sb, params OpCode[] ops)
...
-            sb.EmitPush(CallFlags.All);
-            sb.EmitPush(operation);
-            sb.EmitPush(scriptHash);
-            sb.EmitSysCall(ApplicationEngine.System_Contract_Call);
-            return sb;
+            return EmitDynamicCall(sb, null, scriptHash, operation, args);
         }
 
-        public static ScriptBuilder EmitDynamicCall(this ScriptBuilder sb, UInt160 scriptHash, string operation, params ContractParameter[] args)
+        public static ScriptBuilder EmitDynamicCall(this ScriptBuilder sb, DataCache snapshot, UInt160 scriptHash, string operation, params object[] args)
```
Comment (erikzhang) [inline src/neo/VM/Helper.cs:56]: Is it better to add `CallFlags` as the parameter?

Comment (shargon) [reply src/neo/VM/Helper.cs:56]: optional or two methods? I prefer two methods because the `params`

Comment (erikzhang) [reply src/neo/VM/Helper.cs:56]: Two overloads.

Comment (shargon) [reply src/neo/VM/Helper.cs]: Done

File: src/neo/SmartContract/CallFlags.cs [hunk 2]
```diff
@@ -14,6 +14,7 @@ public enum CallFlags : byte
 
         States = ReadStates | WriteStates,
         ReadOnly = ReadStates | AllowCall,
+        Safe = ReadStates | AllowCall | AllowNotify,
```
Comment (erikzhang) [inline src/neo/SmartContract/CallFlags.cs:17]: I think we don't need `Safe`. We should use `ReadOnly`.

**Review by erikzhang — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Pull request merged**